src/promotion/nhit/nhit.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright(c) 2019 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include "hash.h"
+#include "../../metadata/metadata.h"
+#include "../../ocf_priv.h"
+#include "../../engine/engine_common.h"
+
+#include "nhit.h"
+#include "../ops.h"
+
+#define NHIT_MAPPING_RATIO 2
+
+#define NHIT_MIN_THRESHOLD 1
+#define NHIT_MAX_THRESHOLD 1000
+#define NHIT_THRESHOLD_DEFAULT 3
+
+struct nhit_policy_context {
+	nhit_hash_t hash_map;
+
+	/* Configurable parameters */
+	env_atomic insertion_threshold;
+};
+
+ocf_error_t nhit_init(ocf_cache_t cache, ocf_promotion_policy_t policy)
+{
+	struct nhit_policy_context *ctx;
+	int result = 0;
+
+	ctx = env_vmalloc(sizeof(*ctx));
+	if (!ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+	policy->ctx = ctx;
+
+	result = hash_init(ocf_metadata_get_cachelines_count(cache) *
+			NHIT_MAPPING_RATIO, &ctx->hash_map);
+	if (result)
+		goto dealloc_ctx;
+
+	env_atomic_set(&ctx->insertion_threshold, NHIT_THRESHOLD_DEFAULT);
+
+	return 0;
+
+dealloc_ctx:
+	env_vfree(ctx);
+exit:
+	return result;
+}
+
+void nhit_deinit(ocf_promotion_policy_t policy)
+{
+	struct nhit_policy_context *ctx = policy->ctx;
+
+	hash_deinit(ctx->hash_map);
+
+	env_vfree(ctx);
```

Comment:
After you freed memory, set pointer value to NULL.

Reply:
Fixed